crates/solver/src/settlement_rater.rs

```diff
@@ -244,18 +248,251 @@ impl SettlementRating for SettlementRater {
+                // zero of the payout function lies between `objective - cap`
+                // and `cap`. For such scores the cap in the payment is not
+                // binding, the profit function is just linear in the reference
+                // score and the zero of the function is at
+
+                // `probability_success * score - probability_fail * cost_fail`
+                let score = probability_success * objective.clone() - probability_fail * cost_fail;
+                Ok(score)
+            } else if profit_obj_minus_cap >= zero() && profit_cap > zero() {
+                // optimal score is larger than `objective - cap` and `cap` due to monotonicity
+                // of the profit function. For such reference scores the cap in the payment is
+                // always binding in case of reverts and never in case of success, and the zero
+                // of the function is at
+
+                //score - probability_fail / probability_success * (cap + cost_fail)
+                let score = objective.clone()
+                    - probability_fail
+                        .checked_div(&probability_success)
+                        .context("division by success")?
+                        * (self.score_cap.clone() + cost_fail);
+                Ok(score)
+            } else if profit_obj_minus_cap < zero() && profit_cap <= zero() {
+                // optimal score is smaller than `objective - cap` and `cap` due to monotonicity
+                // of the profit function. In that case, the `cap` is always
+                // binding in case of success and never binds for reverts. Thus
+                // in this range the profit function is again just linear in the
+                // reference score and the zero can be computed as
+
+                // probability_success / probability_fail * cap - cost_fail
+                let score = probability_success
+                    .checked_div(&probability_fail)
+                    .context("division by fail")?
+                    * self.score_cap.clone()
+                    - cost_fail;
+                Ok(score)
+            } else {
+                Err(anyhow!("Invalid bid"))
+            }
+        };
+
+        tracing::trace!(?score, "Optimal score");
+        if let Ok(score) = score.as_ref() {
+            ensure!(score <= &objective, "Optimal score higher than objective");
+        }
+        score
+    }
+
+    // amount that should be payed out to solver or penalized from solver
+    fn profit(
+        &self,
+        score_reference: BigRational,
+        objective: BigRational,
+        probability_success: BigRational,
+        cost_fail: BigRational,
+    ) -> BigRational {
+        tracing::trace!(
+            ?score_reference,
+            ?objective,
+            ?probability_success,
+            ?cost_fail,
+            "Computing profit"
+        );
+
+        // this much is payed out to the solver if the transaction succeeds
```

Comment:
Ok, these comments I find quite obvious from the code (they explain what we do rather than the why we do it)